dashboard/src/actions/profileActions.js

```diff
@@ -0,0 +1,86 @@
+import * as TYPES from "./types";
+import API from "../utils/axiosInstance";
+import * as API_ROUTES from "../utils/apiConstants";
+import { showFailureToast, constructToast } from "./toastActions";
+
+export const getProfileDetails = () => async (dispatch, getState) => {
+  try {
+    dispatch({ type: TYPES.LOADING });
+    const username = getState().userAuth.loginDetails.username;
+
+    const response = await API.get(`${API_ROUTES.GET_USER}/${username}`);
+
+    if (response.status === 200 && Object.keys(response.data).length > 0) {
+      dispatch({
+        type: TYPES.GET_USER_DETAILS,
+        payload: response?.data,
+      });
+    } else {
+      dispatch(showFailureToast());
+    }
+    dispatch({ type: TYPES.COMPLETED });
+  } catch (error) {
+    dispatch(constructToast("danger", error?.response?.data?.message));
+    dispatch({ type: TYPES.NETWORK_ERROR });
+    dispatch({ type: TYPES.COMPLETED });
+  }
+};
+
+export const updateUserDetails =
+  (value, fieldName) => async (dispatch, getState) => {
+    let userDetails = { ...getState().userProfile.userDetails };
+    let updatedUserDetails = { ...getState().userProfile.updatedUserDetails };
+
+    userDetails[fieldName] = value;
+    updatedUserDetails[fieldName] = value;
+    let payload = {
+      userDetails,
+      updatedUserDetails,
+    };
+    dispatch({
+      type: TYPES.UPDATE_USER_DETAILS,
+      payload,
+    });
+  };
+
+export const sendForUpdate = () => async (dispatch, getState) => {
+  try {
+    dispatch({ type: TYPES.LOADING });
+    const updatedUserDetails = { ...getState().userProfile.updatedUserDetails };
+
+    const username = getState().userAuth.loginDetails.username;
+    if (username) {
+      const response = await API.put(`${API_ROUTES.GET_USER}/${username}`, {
+        ...updatedUserDetails,
+      });
+      if (response.status === 200) {
+        dispatch(constructToast("success", "Updated!"));
+        dispatch({
+          type: TYPES.GET_USER_DETAILS,
+          payload: response?.data,
+        });
+        dispatch({ type: TYPES.RESET_DATA });
+        dispatch({ type: TYPES.COMPLETED });
+        return false;
+      } else {
+        dispatch(showFailureToast());
+        dispatch({ type: TYPES.COMPLETED });
+        return false;
+      }
+    }
+  } catch (error) {
+    dispatch(constructToast("danger", error?.response?.data?.message));
+    dispatch({ type: TYPES.NETWORK_ERROR });
+    dispatch({ type: TYPES.COMPLETED });
+  }
+};
+
+export const resetUserDetails = () => async (dispatch, getState) => {
+  const userDetails_copy = { ...getState().userProfile.userDetails_copy };
+
+  dispatch({
+    type: TYPES.SET_USER_DETAILS,
+    payload: { ...userDetails_copy },
+  });
+  dispatch({ type: TYPES.RESET_DATA });
+};
```

Comment:
`userDetails_copy` is a copy of the object returned by `getState().userProfile.userDetails_copy` (which is itself, a copy, I think); we then make a copy of that copy for the `payload` attribute of the object that we pass the first call to `dispatch()`.  Do we really need to make all these copies?  That is, would this work?
```
export const resetUserDetails = () => async (dispatch, getState) => {
 dispatch({
    type: TYPES.SET_USER_DETAILS,
    payload: getState().userProfile.userDetails_copy,
  });
  dispatch({ type: TYPES.RESET_DATA });
};
```
That is, instead of making `payload` refer to a temporary object, make it refer to the original object.

Reply:
Done!